### PR TITLE
[WIP] [aspect_selector] Add test plan

### PR DESCRIPTION
**Pull request by ProGTX**
Initial test plan for the `aspect_selector` feature.

### CODE REVIEW

File: test_plans/SYCL-2020/aspect_selector.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+The CTS does not have to test all combinations of 3 or more aspects.
+
+The CTS will test all the different numbers of aspects at least once, up to `N`,
+where the aspects are not duplicated,
+i.e. `aspect_selector(asp1), aspect_selector(asp1, asp2), ..., aspect_selector(asp1, asp2, ..., aspN)`.
+
+Additional testing will include 100 combinations
+of randomly selected aspect combinations,
+allowing aspect duplicates,
+with a random arity.
```
Comment (gmlueck) [inline test_plans/SYCL-2020/aspect_selector.md:47]: We had an internal discussion recently about adding tests with randomized inputs to the CTS, and we thought it was not a good idea.  We agree that such tests are useful, but we're not sure they belong in the CTS.  Vendors use the CTS to demonstrate conformance, so it seems potentially unfair for the CTS to behave differently for one vendor vs. another.  Consider two vendors that have the same bug.  One might run the CTS and get a pass while another runs the CTS and gets a fail.  That seems unfair.

File: test_plans/SYCL-2020/aspect_selector.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    1. Construct `selectorVector` object obtained by calling `aspect_selector(std::vector{asp1, ..., aspN})`.
+    2. Try/catch block
+
+        1. Construct SYCL device (`selectedDevice`) using `selectorVector`.
+        2. If exception is throw:
+
+            * If `testDevice` contains all the requested aspects,
+              but no device was found using the aspect selector,
+              this is a failed test.
+            * Otherwise there is most likely no device in the system with the requested aspects, skip test.
```
Comment (gmlueck) [inline test_plans/SYCL-2020/aspect_selector.md:83]: Why not iterate over all the devices in the system and see if there is one with all the requested aspects?

File: test_plans/SYCL-2020/aspect_selector.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        3. If test passed so far, there should be no exceptions anymore with further testing.
+        4. The same checks for `testPlatform` (SYCL platform from aspect selector object)
+        and `testQueue` (SYCL queue from aspect selector object).
+
+            * Return `selectedPlatform` and `selectedQueue`, respectively.
+    3. Check that `selectedDevice` contains all requested aspects.
+
+        * `selectedDevice.has(asp1), ..., selectedDevice.has(aspN)`.
+        * The same checks for `selectedPlatform` and `selectedQueue`.
```
Comment (gmlueck) [inline test_plans/SYCL-2020/aspect_selector.md:92]: I think the test for `selectedPlatform` is not be correct.  Let's say the test does this:

```
auto selectVector = aspect_selector(std::vector{asp1});
platform p{selectVector};
```

There is **no** guarantee that `p.has(asp1)` will be true.  The `selectVector` will select some device `d` which has that aspect.  However, the platform will only have that aspect is **all** of its devices have it.  If there is some other device `d2` in that platform which does not have `asp1`, then `p.has(asp1)` will be false.

Note that there is no member function `queue::has()`.  I guess you will test `selectedQueue.get_device().has()`?

**Review by gmlueck — COMMENTED** [review]
I think there should also be some tests for the `denyList` parameter to the overload that takes two  `std::vector` parameters.

File: test_plans/SYCL-2020/aspect_selector.md (hunk 1 above)
Comment (keryell) [reply test_plans/SYCL-2020/aspect_selector.md:47]: On the other hand we can imagine a test-driven implementation of SYCL implementing *only* the CTS... ;-)

File: test_plans/SYCL-2020/aspect_selector.md (hunk 3 above)
Comment (keryell) [reply test_plans/SYCL-2020/aspect_selector.md:92]: So, testing for these behaviors?

File: test_plans/SYCL-2020/aspect_selector.md (hunk 1 above)
Comment (ProGTX) [reply test_plans/SYCL-2020/aspect_selector.md:47]: How about requiring that the generated tests must always produce the same combinations (by using a predefined seed)? I've updated that in the latest wording.

File: test_plans/SYCL-2020/aspect_selector.md (hunk 2 above)
Comment (ProGTX) [reply test_plans/SYCL-2020/aspect_selector.md]: Changed it to that, separated the platform.

File: test_plans/SYCL-2020/aspect_selector.md (hunk 3 above)
Comment (ProGTX) [reply test_plans/SYCL-2020/aspect_selector.md]: You're correct, the latest wording should reflect that.

File: test_plans/SYCL-2020/aspect_selector.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+Assume requested aspects `{asp1, ..., aspN}` for each test,
+and forbidden aspects `denyList = {noasp1, ..., noaspM}`.
+
+These steps are performed before testing combinations:
+
+1. Iterate over all devices to find out if there's one that contains all requested aspects
+    but doesn't contain any forbidden ones,
+    we call this `suitableDevice`.
+
+    * If a device like this cannot be found, skip test.
```
Comment (gmlueck) [inline test_plans/SYCL-2020/aspect_selector.md:89]: Why should the test be skipped in this case?  Instead, I think you could verify that the selector throws an `errc::runtime` exception as the spec mandates.

File: test_plans/SYCL-2020/aspect_selector.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    1. Check that `selectedDevice` contains all requested aspects
+        and none of the forbidden ones.
+
+        * `selectedDevice.has(asp1), ..., selectedDevice.has(aspN)`.
+        * `!selectedDevice.has(noasp1), ..., !selectedDevice.has(noaspN)`.
+
+            * If `denyList` is not empty.
+        * The same checks for `selectedQueue.get_device()`.
+    1. If `suitablePlatform` was found, construct SYCL platform (`selectedPlatform`) using `selectorVector`.
+    1. Check that `selectedPlatform` contains all requested aspects.
```
Comment (gmlueck) [inline test_plans/SYCL-2020/aspect_selector.md:114]: I think this is not correct.  Here is the definition of the platform constructor you are testing:

```
template <typename DeviceSelector>
explicit platform(const DeviceSelector &deviceSelector);
```

> Constructs a SYCL platform instance that is a copy of the platform which **contains the device** returned by the device selector parameter.

(Emphasis mine.)

The returned platform is merely guaranteed to contain **a** device that has all the requested aspects.  However, this does not mean that the platform itself has all of those aspects.  This is because `platform::has(asp)` is only `true` if **all** devices in the platform have aspect `asp`.

Note that this also affects your pre-condition where you check `suitablePlatform`.

File: test_plans/SYCL-2020/aspect_selector.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+For each of the combinations described so far,
+add new combinations which also use forbidden aspects `{noasp1, ..., noaspM}` -
+`M` is independent of `N`, can be smaller, same size, or larger.
+These forbidden aspects will be passed to the `denyList` argument.
+Use the same rules as above for generating aspects,
+following these extra restrictions:
+
+* `noaspY` is not a member of the set `{asp1, ..., aspN}`
+    for each `noaspY` from the set of forbidden aspects.
```
Comment (gmlueck) [inline test_plans/SYCL-2020/aspect_selector.md:60]: I think this is your intent, but you should make it clear that we will also test the case where the `denyList` is empty.

File: test_plans/SYCL-2020/aspect_selector.md (hunk 1 above)
Comment (gmlueck) [reply test_plans/SYCL-2020/aspect_selector.md:47]: Yes, I think this is reasonable.  As long at the test runs the same for each vendor, I'm happy.

**Comment by fraggamuffin** [comment]
contractor?

**Comment by bader** [comment]
https://github.com/KhronosGroup/SYCL-CTS/pull/422 is merged, so I think we can close this one. @ProGTX, are you okay with that?

**Comment by bader** [comment]
> #422 is merged, so I think we can close this one. @ProGTX, are you okay with that?

@ProGTX, any objections to close this PR?

**Comment by ProGTX** [comment]
Closing this PR since it's been replaced.
